src/libraries/System.Private.CoreLib/src/System/Collections/Generic/Dictionary.cs

```diff
@@ -90,25 +90,81 @@ public Dictionary(int capacity, IEqualityComparer<TKey>? comparer)
                 ThrowHelper.ThrowArgumentNullException(ExceptionArgument.dictionary);
             }
 
+            if (dictionary.Count == 0)
+            {
+                // Nothing to copy, all done
+                return;
+            }
+
+            Debug.Assert(_buckets is not null);
+            Debug.Assert(_entries is not null);
+
             // It is likely that the passed-in dictionary is Dictionary<TKey,TValue>. When this is the case,
             // avoid the enumerator allocation and overhead by looping through the entries array directly.
             // We only do this when dictionary is Dictionary<TKey,TValue> and not a subclass, to maintain
             // back-compat with subclasses that may have overridden the enumerator behavior.
             if (dictionary.GetType() == typeof(Dictionary<TKey, TValue>))
             {
                 Dictionary<TKey, TValue> d = (Dictionary<TKey, TValue>)dictionary;
+
+                Debug.Assert(d._buckets is not null);
+                Debug.Assert(d._entries is not null);
+
+                Entry[] oldEntries = d._entries;
+                // If comparers are the same, we can copy dictionary wholesale without rehashing.
+                if (d._comparer == _comparer)
```

Comment:
Is this special case worth the speed up? It creates perf cliff as @GrabYourPitchforks  https://github.com/dotnet/runtime/pull/41944#issuecomment-690813289

(Trying to see whether we can make the code smaller.)

Reply:
Also, these numbers in the table above are interesting. They suggest that the rehashing is faster for large arrays:

Clone_String_Full (30000) is 825.09 us -> 675.24 us
Clone_String_HalfRemoved (30000) is 392.26 -> 194.12 us

The time one would expect for Full with rehasing should be 400 us.

Reply:
The copy only rehashes when `d._comparer != _comparer` so neither the shrinking path nor the array copy paths rehashes and the both use the hashes in the dictionary.

String key is a strange anomaly though as the `.Comparer` property doesn't necessarily give the accurate comparer the Dictionary is actually using for `string`

https://github.com/dotnet/runtime/blob/dcbac854ca05fc0226cfd6ba3e0337e9d429d98e/src/libraries/System.Private.CoreLib/src/System/Collections/Generic/Dictionary.cs#L146-L153

Which means my naive copy using the result of the property may end up with an unequal value and take a slower path
`return new Dictionary<TKey, TValue>(source, source.Comparer);`; however that would go down the `.Add` path which should be the same as the current copy as that's not really changed, so don't think that's what's happening here.

Not sure why the Array.Copy deteriorates faster than the item by item copy though + converting hashcode to bucket index; as it certainly starts faster 🤔

Will rerun and check the asm also

Reply:
Ah; the larger size would go over the `BulkMoveWithWriteBarrierChunk` at 16kB
https://github.com/dotnet/runtime/blob/dcbac854ca05fc0226cfd6ba3e0337e9d429d98e/src/coreclr/src/System.Private.CoreLib/src/System/Buffer.CoreCLR.cs#L31-L40

And then call the `extern` copy
https://github.com/dotnet/runtime/blob/dcbac854ca05fc0226cfd6ba3e0337e9d429d98e/src/coreclr/src/System.Private.CoreLib/src/System/Buffer.CoreCLR.cs#L77-L78

Hence the performance difference?

Reply:
Sorry other way around; too much differentiation by numbers of underscores. It would switch the the C# version from the extern; which then makes multiple calls to the `extern`

https://github.com/dotnet/runtime/blob/dcbac854ca05fc0226cfd6ba3e0337e9d429d98e/src/coreclr/src/System.Private.CoreLib/src/System/Buffer.CoreCLR.cs#L55-L62

Reply:
Not sure anyway around that as `InlinedForwardGCSafeCopyHelper` switches off GC and is doing xmm copies; so not sure could recreate anything faster in C#? (where GC would be active and updating GC for each item copied)

Reply:
I do not think that a few extra calls cannot explain a difference this large for sets of 30,000 items.

Bulk array copying of object references is cheating by being fast in micro-benchmarks by leaving more potential work to be done by the GC. It guess that Benchmark.NET is capturing the extra work done by the GC for larger sets.

Reply:
> Trying to see whether we can make the code smaller.

Aside: would moving it out to a static method in the class make the asm code gen smaller? (less likely to be invoked)